pkg/pf/tests/diff_list_test.go

```diff
@@ -0,0 +1,227 @@
+							Optional: true,
+							PlanModifiers: []planmodifier.String{
+								stringplanmodifier.RequiresReplace(),
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	blockSchema := rschema.Schema{
+		Blocks: map[string]rschema.Block{
+			"key": rschema.ListNestedBlock{
+				NestedObject: rschema.NestedBlockObject{
+					Attributes: map[string]rschema.Attribute{
+						"nested": rschema.StringAttribute{Optional: true},
+					},
+				},
+			},
+		},
+	}
+
+	blockReplaceSchema := rschema.Schema{
+		Blocks: map[string]rschema.Block{
+			"key": rschema.ListNestedBlock{
+				NestedObject: rschema.NestedBlockObject{
+					Attributes: map[string]rschema.Attribute{
+						"nested": rschema.StringAttribute{Optional: true},
+					},
+				},
+				PlanModifiers: []planmodifier.List{
+					listplanmodifier.RequiresReplace(),
+				},
+			},
+		},
+	}
+
+	blockNestedReplaceSchema := rschema.Schema{
+		Blocks: map[string]rschema.Block{
+			"key": rschema.ListNestedBlock{
+				NestedObject: rschema.NestedBlockObject{
+					Attributes: map[string]rschema.Attribute{
+						"nested": rschema.StringAttribute{
+							Optional: true,
+							PlanModifiers: []planmodifier.String{
+								stringplanmodifier.RequiresReplace(),
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	attrList := func(arr *[]string) cty.Value {
+		if arr == nil {
+			return cty.NullVal(cty.DynamicPseudoType)
+		}
+		slice := make([]cty.Value, len(*arr))
+		for i, v := range *arr {
+			slice[i] = cty.StringVal(v)
+		}
+		if len(slice) == 0 {
+			return cty.ListValEmpty(cty.String)
+		}
+		return cty.ListVal(slice)
+	}
+
+	nestedAttrList := func(arr *[]string) cty.Value {
+		if arr == nil {
+			return cty.NullVal(cty.DynamicPseudoType)
+		}
+		slice := make([]cty.Value, len(*arr))
+		for i, v := range *arr {
+			slice[i] = cty.ObjectVal(
+				map[string]cty.Value{
+					"nested": cty.StringVal(v),
+				},
+			)
+		}
+		if len(slice) == 0 {
+			return cty.ListValEmpty(cty.Object(map[string]cty.Type{"nested": cty.String}))
+		}
+		return cty.ListVal(slice)
+	}
+
+	schemaValueMakerPairs := []struct {
+		name       string
+		schema     rschema.Schema
+		valueMaker func(*[]string) cty.Value
+	}{
+		{"attribute no replace", attributeSchema, attrList},
+		{"attribute requires replace", attributeReplaceSchema, attrList},
+		{"nested attribute no replace", nestedAttributeSchema, nestedAttrList},
+		{"nested attribute requires replace", nestedAttributeReplaceSchema, nestedAttrList},
+		{"nested attribute nested requires replace", nestedAttributeNestedReplaceSchema, nestedAttrList},
+		{"block no replace", blockSchema, nestedAttrList},
+		{"block requires replace", blockReplaceSchema, nestedAttrList},
+		{"block nested requires replace", blockNestedReplaceSchema, nestedAttrList},
+	}
+
+	scenarios := []struct {
+		name         string
+		initialValue *[]string
+		changeValue  *[]string
+	}{
+		{"unchanged non-empty", &[]string{"value"}, &[]string{"value"}},
```

Comment:
The three-element scenarios can actually be interesting to see preview. Inserting an element in the middle of 2 elements.